src/plugins/postgres_permissions_operator.py

```diff
@@ -0,0 +1,171 @@
+datapunt_evironment = env("DATAPUNT_ENVIRONMENT")
+
+
+class PostgresPermissionsOperator(BaseOperator):
+    """
+    This operator executes DB grants << on >> DB tables / fields << to >> DB roles.
+
+    It can be envoked based upon a single dataset or a list of datasets. It uses the dag_id to identify the dataset.
+
+    The DB roles are derived from the AUTH element as defined in the Amsterdam schema.
+    When no AUTH element is provided in the Amsterdam schema, it default to DB role SCOPE_OPENBAAR.
+    DB users can be assigned as member to the DB roles to get access to DB objects.
+
+    Caveat:
+    When objects are dropped in the DB, the grants to the roles are removed as well.
+    This results that grants must be (re)applied again to the roles.
+
+    An other way to deal with is, is to update tables based upon the pgcomparator_cdc_operator.py.
+    Which detects differences between source and target table, and upserts the target table if nessacery
+    in stead of dropping, recreating the target table and inserting the records on each run.
+    """
+
+    def __init__(
+        self,
+        batch_ind=False,
+        batch_timewindow="0:30",
+        dag_name=None,
+        db_conn=default_db_conn,
+        schema_url=default_schema_url,
+        db_schema="public",
+        profiles=None,
+        role="AUTO",
+        scope=None,
+        dry_run=False,
+        create_roles=True,
+        revoke=False,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.batch_ind = batch_ind
+        self.batch_timewindow = batch_timewindow
+        self.dataset_name = dag_name
+        self.db_conn = db_conn
+        self.schema_url = schema_url
+        self.db_schema = db_schema
+        self.profiles = profiles
+        self.role = role
+        self.scope = scope
+        self.dry_run = dry_run
+        self.create_roles = create_roles
+        self.revoke = revoke
+
+    def execute(self, context=None):
+        """Executes the 'apply_schema_and_profile_permissions' method from schema-tools to set the database permissions on objects to roles"""
+
+        # setup logger so output can be added to the Airflow logs
+        logger = logging.getLogger(__name__)
+
+        # setup database connection where the database objects are present
+        engine = _get_engine(self.db_conn)
+
+        # Option ONE: batch grant
+        if self.batch_ind:
+
+            # get current datetime and make it aware of the TZ (mandatory)
+            now = pendulum.now("Europe/Amsterdam")
+
+            # calculate the delta between current datetime and specified time window
+            time_window_hour = int(self.batch_timewindow.split(':')[0])
+            time_window_minutes = int(self.batch_timewindow.split(':')[1])
+            delta = now - timedelta(hours=time_window_hour, minutes=time_window_minutes)
+
+            logger.info(f"the time window is set starting at: {delta} till now")
```

Comment:
Bij de logger is het beter om de parameters in de vorm van %s (or %r) te gebruiken. Het zal hier niet veel uitmaken, maar het idee is dat de interpolatie van de string alleen plaatsvindt als het echt nodig is. Is dus vooral relevant als de logging regel binnen een if-statement staat bijv. Maar, het is een goede gewoonte om voor logging de %-interpolatie te gebruiken.

Reply:
Goed te weten, pas ik aan!